File: designer/src/app/core/services/execution-engine.service.ts
import { Injectable } from '@angular/core';
import {
  HttpNodeConfig,
  CodeNodeConfig,
  ConditionNodeConfig,
  DoNothingNodeConfig,
  ExecutionResult,
  NodeRunResult,
  TriggerNodeConfig,
  WorkflowEdge,
  WorkflowNode
} from '../models/workflow.types';
import { ExecutionContext } from './execution-context';
import { topologicalSort } from '../utils/topological-sort';
import { runTriggerNode } from './runners/trigger';
import { runHttpNode } from './runners/http-post';
import { runConditionNode } from './runners/condition';
import { runDoNothingNode } from './runners/nothing';
import { runCodeNode } from './runners/code';
import {
  concatMap,
  defer,
  from,
  map,
  Observable,
  of,
  takeWhile,
  tap,
  toArray
} from 'rxjs';

@Injectable({ providedIn: 'root' })
export class ExecutionEngineService {
  execute(nodes: WorkflowNode[], edges: WorkflowEdge[]): Observable<ExecutionResult> {
    return defer(() => {
      const runId = `run-${Date.now()}-${Math.random().toString(36).slice(2, 9)}`;
      const context = new ExecutionContext(runId);
      const validationError = this.validateWorkflow(nodes);
      if (validationError) {
        throw new Error(validationError);
      }

      const { sorted, hasCycle } = topologicalSort(nodes, edges);
      if (hasCycle) {
        throw new Error('Cycle detected in workflow graph');
      }

      const triggerNode = sorted.find((n: WorkflowNode) => n.type === 'TRIGGER');
      if (!triggerNode) {
        throw new Error('No trigger node found');
      }

      return from(sorted).pipe(
        concatMap(node => {
          const snapshot = context.snapshot();
          const shouldExecute = this.shouldExecuteNode(node, edges, context);

          if (!shouldExecute) {
            const skipped: NodeRunResult = {
              nodeId: node.id,
              outputs: {},
              status: 'skipped',
              timestamp: new Date().toISOString()
            };
            context.addNodeResult(skipped);
            return of(skipped);
          }

          let runner$: Observable<NodeRunResult>;
          switch (node.type) {
            case 'TRIGGER':
              runner$ = runTriggerNode(node.id, node.data.config as TriggerNodeConfig, snapshot);
              break;
            case 'CIBIL':
            case 'CRIF':
            case 'EXPERIAN':
            case 'EQUIFIX':
              runner$ = runHttpNode(node.id, node.data.config as HttpNodeConfig, snapshot);
              break;
            case 'CONDITION':
              runner$ = runConditionNode(node.id, node.data.config as ConditionNodeConfig, snapshot);
              break;
            case 'DO_NOTHING':
              runner$ = runDoNothingNode(node.id, node.data.config as DoNothingNodeConfig, snapshot);
              break;
            case 'CODE':
              runner$ = runCodeNode(node.id, node.data.config as CodeNodeConfig, snapshot);
              break;
            default:
              runner$ = of({
                nodeId: node.id,
                outputs: {},
                status: 'failed',
                error: `Unknown node type: ${node.type}`,
                timestamp: new Date().toISOString()
              });
          }

          return runner$.pipe(
            tap((result: NodeRunResult) => context.addNodeResult(result)),
            takeWhile((result: NodeRunResult) => result.status !== 'failed', true) // include failing result then stop
          );
        }),
        toArray(),
        map(results => ({
          runId,
          results
        }))
      );
    });
  }

  private shouldExecuteNode(
    node: WorkflowNode,
    edges: WorkflowEdge[],
    context: ExecutionContext
  ): boolean {
    const incomingEdges = edges.filter(e => e.target === node.id);
    if (incomingEdges.length === 0) return true;

    for (const edge of incomingEdges) {
      const sourceOutputs = context.nodeOutputs[edge.source];
      if (!sourceOutputs) return false;

      if (Object.prototype.hasOwnProperty.call(sourceOutputs, 'branch')) {
        const expectedBranch = edge.sourceHandle ?? 'true';
        if ((sourceOutputs as Record<string, unknown>)['branch'] !== expectedBranch) {
          return false;
        }
      }
    }

    return true;
  }

  private validateWorkflow(nodes: WorkflowNode[]): string | null {
    const triggerNodes = nodes.filter((n: WorkflowNode) => n.type === 'TRIGGER');
    if (triggerNodes.length === 0) {
      return 'Workflow must have at least one Trigger node';
    }

    for (const node of nodes) {
      if (node.type === 'CIBIL' || node.type === 'CRIF' || node.type === 'EXPERIAN' || node.type === 'EQUIFIX') {
        const config = node.data.config as HttpNodeConfig;
        if (!config.url || config.url.trim() === '') {
          return `${node.type} node ${node.id} must have a URL`;
        }
      }
    }

    for (const node of nodes) {
      if (node.type === 'CONDITION') {
        const config = node.data.config as ConditionNodeConfig;
        if (!config.expression || config.expression.trim() === '') {
          return `Condition node ${node.id} must have an expression`;
        }
        try {
          new Function('snapshot', `return ${config.expression}`);
        } catch (error) {
          return `Condition node ${node.id} has invalid JS syntax: ${error}`;
        }
      }
    }

    return null;
  }
}

File: api/src/main/java/com/workflow/api/service/ExecutionEngineService.java
package com.workflow.api.service;

import com.workflow.api.dto.*;
import com.workflow.api.service.runners.*;
import com.workflow.api.util.TopologicalSort;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import java.time.Instant;
import java.util.*;
import java.util.stream.Collectors;

@Slf4j
@Service
@RequiredArgsConstructor
public class ExecutionEngineService {
    
    private final TriggerNodeRunner triggerNodeRunner;
    private final HttpNodeRunner httpNodeRunner;
    private final ConditionNodeRunner conditionNodeRunner;
    private final DoNothingNodeRunner doNothingNodeRunner;
    private final CodeNodeRunner codeNodeRunner;
    
    public ExecutionResult execute(List<WorkflowNode> nodes, List<WorkflowEdge> edges) {
        String runId = "run-" + System.currentTimeMillis() + "-" + UUID.randomUUID().toString().substring(0, 8);
        ExecutionContext context = new ExecutionContext(runId);
        
        // Validate workflow
        String validationError = validateWorkflow(nodes);
        if (validationError != null) {
            throw new IllegalArgumentException(validationError);
        }
        
        // Topological sort
        TopologicalSort.Result sortResult = TopologicalSort.sort(nodes, edges);
        if (sortResult.isHasCycle()) {
            throw new IllegalArgumentException("Cycle detected in workflow graph");
        }
        
        List<WorkflowNode> sorted = sortResult.getSorted();
        
        // Find trigger node
        WorkflowNode triggerNode = sorted.stream()
            .filter(n -> "TRIGGER".equals(n.getType()))
            .findFirst()
            .orElseThrow(() -> new IllegalArgumentException("No trigger node found"));
        
        List<NodeRunResult> results = new ArrayList<>();
        
        // Execute nodes in topological order
        for (WorkflowNode node : sorted) {
            ExecutionSnapshot snapshot = context.snapshot();
            
            // Check if node should execute
            if (!shouldExecuteNode(node, edges, context)) {
                NodeRunResult skipped = new NodeRunResult(
                    node.getId(),
                    new HashMap<>(),
                    "skipped",
                    null,
                    Instant.now().toString()
                );
                context.addNodeResult(skipped);
                results.add(skipped);
                continue;
            }
            
            // Execute node
            NodeRunResult result = executeNode(node, snapshot);
            context.addNodeResult(result);
            results.add(result);
            
            // Stop execution if node failed
            if ("failed".equals(result.getStatus())) {
                break;
            }
        }
        
        return new ExecutionResult(runId, results);
    }
    
    public NodeRunResult executeSingleNode(WorkflowNode node) {
        ExecutionContext context = new ExecutionContext("test-" + System.currentTimeMillis());
        ExecutionSnapshot snapshot = context.snapshot();
        return executeNode(node, snapshot);
    }
    
    private NodeRunResult executeNode(WorkflowNode node, ExecutionSnapshot snapshot) {
        String nodeType = node.getType();
        
        return switch (nodeType) {
            case "TRIGGER" -> triggerNodeRunner.run(node, snapshot);
            case "HTTP_REQUEST" -> httpNodeRunner.run(node, snapshot);
            case "CONDITION" -> conditionNodeRunner.run(node, snapshot);
            case "DO_NOTHING" -> doNothingNodeRunner.run(node, snapshot);
            case "CODE" -> codeNodeRunner.run(node, snapshot);
            default -> new NodeRunResult(
                node.getId(),
                new HashMap<>(),
                "failed",
                "Unknown node type: " + nodeType,
                Instant.now().toString()
            );
        };
    }
    
    private boolean shouldExecuteNode(WorkflowNode node, List<WorkflowEdge> edges, ExecutionContext context) {
        List<WorkflowEdge> incomingEdges = edges.stream()
            .filter(e -> e.getTarget().equals(node.getId()))
            .collect(Collectors.toList());
        
        if (incomingEdges.isEmpty()) {
            return true;
        }
        
        for (WorkflowEdge edge : incomingEdges) {
            Map<String, Object> sourceOutputs = context.getNodeOutputs().get(edge.getSource());
            if (sourceOutputs == null) {
                return false;
            }
            
            // Check for conditional branching
            if (sourceOutputs.containsKey("branch")) {
                String expectedBranch = edge.getSourceHandle() != null ? edge.getSourceHandle() : "true";
                Object actualBranch = sourceOutputs.get("branch");
                if (!expectedBranch.equals(String.valueOf(actualBranch))) {
                    return false;
                }
            }
        }
        
        return true;
    }
    
    private String validateWorkflow(List<WorkflowNode> nodes) {
        long triggerCount = nodes.stream()
            .filter(n -> "TRIGGER".equals(n.getType()))
            .count();
        
        if (triggerCount == 0) {
            return "Workflow must have at least one Trigger node";
        }
        
        for (WorkflowNode node : nodes) {
            String nodeType = node.getType();
            if ("HTTP_REQUEST".equals(nodeType)) {
                Map<String, Object> config = node.getData().getConfig();
                String url = (String) config.get("url");
                if (url == null || url.trim().isEmpty()) {
                    return "HTTP_REQUEST node " + node.getId() + " must have a URL";
                }
            }
            
            if ("CONDITION".equals(nodeType)) {
                Map<String, Object> config = node.getData().getConfig();
                String expression = (String) config.get("expression");
                if (expression == null || expression.trim().isEmpty()) {
                    return "Condition node " + node.getId() + " must have an expression";
                }
            }
        }
        
        return null;
    }
}
